pkg/cloud-init/cloud-init.go

```diff
@@ -27,39 +27,155 @@ import (
 	"os/exec"
 	"time"
 
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	v1 "kubevirt.io/kubevirt/pkg/api/v1"
 	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
 	"kubevirt.io/kubevirt/pkg/kubecli"
 	"kubevirt.io/kubevirt/pkg/log"
 	"kubevirt.io/kubevirt/pkg/precond"
+	"kubevirt.io/kubevirt/pkg/util/net/dns"
 )
 
-type IsoCreationFunc func(isoOutFile string, inFiles []string) error
+type IsoCreationFunc func(isoOutFile, volumeID string, inDir string) error
 
 var cloudInitLocalDir = "/var/run/libvirt/cloud-init-dir"
 var cloudInitOwner = "qemu"
 var cloudInitIsoFunc = defaultIsoFunc
 
-const NoCloudFile = "noCloud.iso"
+// Locations of data source disk files
+const (
+	noCloudFile     = "noCloud.iso"
+	configDriveFile = "configdrive.iso"
+)
+
+type DataSourceType string
 
-// Supported DataSources
 const (
-	dataSourceNoCloud = "noCloud"
+	DataSourceNoCloud     DataSourceType = "noCloud"
+	DataSourceConfigDrive DataSourceType = "configDrive"
 )
 
-func defaultIsoFunc(isoOutFile string, inFiles []string) error {
+// CloudInitData is a data source independent struct that
+// holds cloud-init user and network data
+type CloudInitData struct {
+	DataSource  DataSourceType
+	MetaData    string
+	UserData    string
+	NetworkData string
+}
+
+// IsValidCloudInitData checks if the given CloudInitData object is valid in the sense that GenerateLocalData can be called with it.
+func IsValidCloudInitData(cloudInitData *CloudInitData) bool {
+	return cloudInitData != nil && cloudInitData.UserData != "" && cloudInitData.MetaData != ""
+}
+
+// ReadCloudInitVolumeDataSource scans the given VMI for CloutInit volumes and
+// reads their content into a CloudInitData struct. Does not resolve secret refs.
+// To ensure that secrets are read correctly, call InjectCloudInitSecrets beforehand.
+func ReadCloudInitVolumeDataSource(vmi *v1.VirtualMachineInstance) (cloudInitData *CloudInitData, err error) {
+	precond.MustNotBeNil(vmi)
+	hostname := dns.SanitizeHostname(vmi)
+
+	for _, volume := range vmi.Spec.Volumes {
+		if volume.CloudInitNoCloud != nil {
+			cloudInitData, err = readCloudInitNoCloudSource(volume.CloudInitNoCloud)
+			cloudInitData.MetaData = readCloudInitNoCloudMetaData(vmi.Name, hostname, vmi.Namespace)
+			return cloudInitData, err
+		}
+		if volume.CloudInitConfigDrive != nil {
+			cloudInitData, err = readCloudInitConfigDriveSource(volume.CloudInitConfigDrive)
+			cloudInitData.MetaData = readCloudInitConfigDriveMetaData(string(vmi.UID), vmi.Name, hostname, vmi.Namespace)
+			return cloudInitData, err
+		}
+	}
+	return nil, nil
+}
+
+func readCloudInitNoCloudSource(source *v1.CloudInitNoCloudSource) (*CloudInitData, error) {
```

Comment:
`readCloudInitNoCloudSource` and `readCloudInitConfigDriveSource` are almost the same, could you refactor these two functions as well? For example by extracting common parts to new smaller functions, or by creating generic one with type assertion?

Reply:
I tried to merge as much common code as possible into new common functions.
I guess if we want to merge even more code it would be necessary to pick up @greg-bock's suggestion from above and merge the NoCloud and ConfigDrive volume sources into a generic CloudInit volume source.

Reply:
Okay then, for this PR I am okay to leave it like that.